enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,321 @@
+---
+title: gcp_user_defined_labels
+authors:
+  - "@bhb"
+reviewers:
+  - "@patrickdillon" ## reviewer for installer component
+  - "@JoelSpeed" ## reviewer for api and machine-api-provider-gcp components
+  - "@dmage" ## reviewer for cluster-image-registry-operator component
+  - "@akhil-rane" ## reviewer for cloud-credential-operator component
+approvers:
+  - "@jerpeter1" ## approver for CFE
+api-approvers:
+  - "@JoelSpeed" ## approver for api component
+creation-date: 2022-07-12
+last-updated: 2022-07-12
+tracking-link:
+  - https://issues.redhat.com/browse/OCPPLAN-8155
+  - https://issues.redhat.com/browse/CORS-2249
+see-also:
+  - "enhancements/api-review/custom-tags-aws.md"
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Apply user defined labels to all GCP resources created by OpenShift
+
+## Summary
+
+This enhancement describes the proposal to allow an administrator of OpenShift to 
+have the ability to apply user defined labels to those resources created by OpenShift 
+in GCP.
+
+## Motivation
+
+Motivations include but are not limited to:
+
+- Allow admin, compliance, and security teams to keep track of assets and objects 
+  created by OpenShift in GCP.
+
+### User Stories
+
+- As an openshift administrator, I want to have labels added to all resources created 
+  in GCP by Openshift.
+
+### Goals
+
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
+  can configure allowed number of user-defined labels in the Openshift installer generated
+  install config, which is referenced and applied by installer and the in-cluster operators
+  on the the GCP resources during creation.
+- Labels must be applied at creation time, in an atomic operation. It isn't acceptable 
+  to create an object and to apply labels post cluster creation.
+
+### Non-Goals
+
+- Management(update/delete) of resource labels post creation of cluster is out of scope.
+
+## Proposal
+
+A label of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
+resource created by Openshift to enable administrator to differentiate the resources
+created for Openshift cluster. An administrator is not allowed to add or modify the label 
+having the prefix `kubernetes.io` or `openshift.io` in the name. 
+
+New `userLabels` field will be added to `platform.gcp` of install-config for the user 
+to define the labels to be added to the resources created by installer and in-cluster operators.
+
+If `platform.gcp.userLabels` of install-config has any label defined, same will be added 
+to all the gcp resources created by Openshift except, when the label validation fails 
+to adhere to any of the below conditions
+1. A label name must have minimum of 1 character and can have maximum of 63 characters.
+2. A label value can have maximum of 63 characters.
```

Comment:
The string `kubernetes.io/cluster/` is 22 characters, which leave 41 characters for the cluster name, what happens if the cluster name exceeds 41 characters?

Reply:
Cluster name [generated](https://github.com/openshift/api/blob/28228c0f01ab20e1d49d8369157fb171ad53e102/config/v1/types_infrastructure.go#L52-L55) by appending a five characters length random string(+1 for `-` used as delimiter) to the user inputted name is restricted to 27 characters, which is within the limit and is not a problem I think.
/cc @patrickdillon Could you please confirm, if my understanding is correct?

Reply:
The installer team already is currently handling this (for a different problem) on destroy here: https://github.com/barbacbd/installer/blob/master/pkg/destroy/gcp/disk.go#L23